File: content-solutions/dfs/maze-escape.java
class Solution {
    private String[] maze;
    private boolean[][] visited;
    private int rows, cols;
    private static final int[][] DIRS = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    void dfs(int r, int c) {
        visited[r][c] = true;
        for (int[] d : DIRS) {
            int nr = r + d[0], nc = c + d[1];
            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && maze[nr].charAt(nc) != '#' && !visited[nr][nc]) {
                dfs(nr, nc);
            }
        }
    }

    public boolean solution(String[] maze) {
        this.maze = maze;
        rows = maze.length;
        cols = maze[0].length();
        visited = new boolean[rows][cols];
        for (int r = 0; r < rows; r++)
            for (int c = 0; c < cols; c++)
                if (maze[r].charAt(c) == 'S') dfs(r, c);
        for (int r = 0; r < rows; r++)
            for (int c = 0; c < cols; c++)
                if (maze[r].charAt(c) == 'E') return visited[r][c];
        return false;
    }
}

File: content-solutions/dfs/maze-escape.js
function solution(maze) {
  const rows = maze.length;
  const cols = maze[0].length;
  const visited = Array.from({ length: rows }, () => new Array(cols).fill(false));
  const dirs = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];
  function dfs(r, c) {
    visited[r][c] = true;
    for (const [dr, dc] of dirs) {
      const nr = r + dr;
      const nc = c + dc;
      if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && maze[nr][nc] !== "#" && !visited[nr][nc]) dfs(nr, nc);
    }
  }
  for (let r = 0; r < rows; r += 1) {
    for (let c = 0; c < cols; c += 1) {
      if (maze[r][c] === "S") dfs(r, c);
    }
  }
  for (let r = 0; r < rows; r += 1) {
    for (let c = 0; c < cols; c += 1) {
      if (maze[r][c] === "E") return visited[r][c];
    }
  }
  return false;
}
